File: day1.py
import aoc

        
def solve(lines: list[str]):
    """Solves part 1"""
    left = []
    right = []
    for line in lines:
        num1, num2 = line.split()
        left.append(int(num1))
        right.append(int(num2))

    left.sort()
    right.sort()

    sum = 0
    for l, r in zip(left, right):
        sum += abs(l - r)

    return sum


def solve2(lines: list[str]):
    left = []
    right = {}
    for line in lines:
        l, r = line.split()
        left.append(int(l))
        r = int(r)
        if r in right:
            right[r] += 1
        else:
            right[r] = 1

    sum = 0
    for val in left:
        try:
            sum += val * right[val]
        except KeyError:
            continue

    return sum


if __name__ == '__main__':
    print(solve(aoc.get_input_lines('day1.txt')))

File: day1.hs
import System.IO
import Data.List


getFirstList :: [[Int]] -> [Int]
getFirstList = map head

getSecondList :: [[Int]] -> [Int]
getSecondList = map (!! 1)

getNums :: String -> [Int]
getNums = map read . words

absDiff :: Int -> Int -> Int
absDiff a b = abs $ a - b

solve :: String -> Int
solve inp = sum $ zipWith absDiff lst1 lst2
    where
        lst1 = sort $ getFirstList $ map getNums $ lines inp
        lst2 = sort $ getSecondList $ map getNums $ lines inp

appears :: [Int] -> Int -> Int
appears lst n = length $ filter (== n) lst

solve' :: String -> Int
solve' inp = foldr (f lst2) 0 lst1
    where
        lst1 = sort $ getFirstList $ map getNums $ lines inp
        lst2 = sort $ getSecondList $ map getNums $ lines inp
        f :: [Int] -> Int -> Int -> Int
        f lst new buildUp = buildUp + new * appears lst new

main :: IO ()
main = do
    input <- readFile "day1.txt"

    let res = solve' input

    print res